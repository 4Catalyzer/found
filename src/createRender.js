import React from 'react';

import ElementsRenderer from './ElementsRenderer';

export default function createRender({ renderError }) {
  // This is not a component.
  // eslint-disable-next-line react/prop-types
  return function render({ error, elements }) {
    if (error) {
      return renderError ? renderError(error) : null;
    }

    return <ElementsRenderer elements={elements} />;
  };
}
