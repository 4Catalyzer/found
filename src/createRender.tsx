import React from 'react';

import ElementsRenderer, {
  RenderArgs,
  RenderErrorArgs,
  RenderPendingArgs,
  RenderReadyArgs,
} from './ElementsRenderer';
import StaticContainer from './StaticContainer';

export interface CreateRenderOptions {
  renderPending?: (args: RenderPendingArgs) => React.ReactElement;
  renderReady?: (args: RenderReadyArgs) => React.ReactElement;
  renderError?: (args: RenderErrorArgs) => React.ReactNode;
}

/**
 * A convenience method for handling the 3 main states a route match might produce.
 */
export default function createRender({
  renderPending,
  renderReady,
  renderError,
}: CreateRenderOptions): (renderArgs: RenderArgs) => React.ReactElement {
  return function render(renderArgs: RenderArgs): React.ReactElement {
    // TODO: error and elements MIGHT exist, depending on type of renderArgs
    const { error, elements } = renderArgs as any;
    let element;

    if (error) {
      element = renderError ? renderError(renderArgs as any) : null;
    } else if (!elements) {
      element = renderPending ? renderPending(renderArgs) : undefined;
    } else if (renderReady) {
      element = renderReady(renderArgs as any);
    } else {
      element = <ElementsRenderer elements={elements} />;
    }

    const hasElement = element !== undefined;

    return (
      <StaticContainer shouldUpdate={hasElement}>
        {hasElement ? element : null}
      </StaticContainer>
    );
  };
}
