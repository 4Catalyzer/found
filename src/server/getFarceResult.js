import FarceActions from 'farce/lib/Actions';
import ServerProtocol from 'farce/lib/ServerProtocol';
import React from 'react';
import { Provider } from 'react-redux';

import getStoreRenderArgs from '../getStoreRenderArgs';
import RedirectException from '../RedirectException';
import defaultResolver from '../resolver';
import createFarceStore from '../utils/createFarceStore';

export default async function getFarceResult({
  url,
  historyMiddlewares,
  historyOptions,
  routeConfig,
  matchContext,
  resolver = defaultResolver,
  render,
}) {
  const store = createFarceStore({
    historyProtocol: new ServerProtocol(url),
    historyMiddlewares,
    historyOptions,
    routeConfig,
  });

  let renderArgs;

  try {
    renderArgs = await getStoreRenderArgs({
      store,
      matchContext,
      resolver,
    });
  } catch (e) {
    if (e instanceof RedirectException) {
      // The store is not exposed to the user, so we need to build the redirect
      // URL here.
      return {
        redirect: {
          url: store.farce.createHref(e.location),
        },
      };
    }

    throw e;
  } finally {
    // This is a no-op with ServerProtocol, but it doesn't hurt.
    store.dispatch(FarceActions.dispose());
  }

  return {
    status: renderArgs.error ? renderArgs.error.status : 200,
    element: <Provider store={store}>{render(renderArgs)}</Provider>,
  };
}
