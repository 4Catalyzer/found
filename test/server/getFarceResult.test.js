import React from 'react';
import ReactDOMServer from 'react-dom/server';

import createRender from '../../src/createRender';
import getFarceResult from '../../src/server/getFarceResult';

async function renderToString(url, routeConfig) {
  const { element } = await getFarceResult({
    url,
    routeConfig,
    render: createRender({}),
  });

  return ReactDOMServer.renderToString(element);
}

describe('getFarceResult', () => {
  it('should support nested routes', async () => {
    expect(
      await renderToString('/foo/baz/a', [
        {
          path: 'foo',
          Component: ({ children }) => <div className="foo">{children}</div>,
          children: [
            {
              path: 'bar',
              Component: () => <div className="bar" />,
            },
            {
              path: 'baz/:qux',
              Component: ({ match }) => (
                <div className="baz">{match.params.qux}</div>
              ),
            },
          ],
        },
      ]),
    ).toMatchInlineSnapshot(
      `"<div class=\\"foo\\"><div class=\\"baz\\">a</div></div>"`,
    );
  });

  it('should support named child routes', async () => {
    expect(
      await renderToString('/foo/bar/qux/a', [
        {
          path: 'foo',
          Component: ({ nav, main }) => (
            <div className="foo">
              {nav}
              {main}
            </div>
          ),
          children: [
            {
              path: 'bar',
              children: {
                nav: [
                  {
                    path: '(.*)?',
                    Component: () => <div className="bar-nav" />,
                  },
                ],
                main: [
                  {
                    path: 'baz',
                    Component: () => <div className="baz" />,
                  },
                  {
                    path: 'qux/:quux',
                    Component: ({ match }) => (
                      <div className="qux">{match.params.quux}</div>
                    ),
                  },
                ],
              },
            },
          ],
        },
      ]),
    ).toMatchInlineSnapshot(
      `"<div class=\\"foo\\"><div class=\\"bar-nav\\"></div><div class=\\"qux\\">a</div></div>"`,
    );
  });
});
