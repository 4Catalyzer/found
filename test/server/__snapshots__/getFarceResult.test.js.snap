// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFarceResult should support named child routes 1`] = `"<div class=\\"foo\\"><div class=\\"bar-nav\\"></div><div class=\\"qux\\">a</div></div>"`;

exports[`getFarceResult should support nested routes 1`] = `"<div class=\\"foo\\"><div class=\\"baz\\">a</div></div>"`;
